src/thunderbots/software/network_input/filter/ball_filter.h

```diff
@@ -29,21 +58,94 @@ class BallFilter
+     */
+    size_t getAdjustedBufferSize(
+        boost::circular_buffer<SSLBallDetection> ball_detections);
+
+    /**
+     * Adds ball detections to the buffer stored by this filter. This function will ignore
+     * data that is outside of the field, and data that is too far away from the current
+     * known ball position and therefore is likely to be random noise.
+     *
+     * @param new_ball_detections The ball detections to try add to the buffer
+     * @param field The field being played on.
+     */
+    void addNewDetectionsToBuffer(std::vector<SSLBallDetection> new_ball_detections,
+                                  const Field &field);
+
+    /**
+     * Estimates the ball's velocitybased on the current detections in the given buffer.
+     * If the ball_regression_line is provided, the detection positions are projected onto
+     * the line before the velocities are calculated.
+     *
+     * @param ball_detections The ball detections to use to calculate
+     * @param ball_regression_line The ball_regression_line to snap detections to before
+     * calculating velocities.
+     * @return A struct containing various estimates of the ball's velocity based on the
+     * given detections
+     */
+    BallVelocityEstimate estimateBallVelocity(
+        boost::circular_buffer<SSLBallDetection> ball_detections,
+        const std::optional<Line> &ball_regression_line = std::nullopt);
+
+    /**
+     * Given a list of ball detections, use linear regression to find a line of best fit
+     * through the ball positions, and calculate the error of this regression.
+     *
+     * @param ball_detections The ball detections to use in the regression
+     * @return A struct containing the regression line and error of the linear regression
+     */
+    LinearRegressionResults getLinearRegressionLine(
+        boost::circular_buffer<SSLBallDetection> ball_detections);
+
+    /**
+     * Uses linear regression to filter the given list of ball detections to fine the
+     * current "real" state of the ball.
+     *
+     * @param ball_detections The detections to filter
+     * @return The filtered current state of the ball
+     */
+    Ball getLinearRegressionPositionAndVelocity(
+        boost::circular_buffer<SSLBallDetection> ball_detections);
 
    private:
-    boost::circular_buffer<Vector> previous_ball_readings;
+    // A circular buffer used to store previous ball detections, so we can use them
+    // in the filter
+    boost::circular_buffer<SSLBallDetection> ball_detection_buffer;
+    unsigned int _min_buffer_size;
+    unsigned int _max_buffer_size;
+    // If the estimated ball speed is greater than this value, the smallest possible
+    // buffer will be used by the filter
+    double MAX_BUFFER_SIZE_VELOCITY_MAGNITUDE = 4.0;
+    // If the estimated ball speed is less than this value, the largest possible buffer
+    // will be used by the filter
+    double MIN_BUFFER_SIZE_VELOCITY_MAGNITUDE = 0.5;
```

Comment:
Why do we initialize these values here, but `_min_buffer_size` and `_max_buffer_size` elsewhere? Also why are these all not constants?

Reply:
resolved after in person discussion